who_l3_smart_tools/core/terminology/terminology.py

```diff
@@ -0,0 +1,256 @@
+                continue
+            self.converted_row["extras"][
+                self.schema.format_extras_for_json(key)
+            ] = value
+
+
+class ConceptTerminology:
+    """
+    Represents a concept terminology.
+
+    Args:
+        owner_id (str): The ID of the owner.
+        concept_class (str): The concept class.
+        files (list[str]): A list of file paths.
+        schema (type[ConceptSchema], optional): The concept schema type. Defaults to ConceptSchema.
+
+    Attributes:
+        files (list[str]): A list of csv file paths.
+        schema (type[ConceptSchema]): The concept schema type.
+        output_csv_header (list[str]): The CSV header for the output file.
+        extra_args (dict): Extra arguments for processing the concept.
+
+    """
+
+    def __init__(
+        self,
+        owner_id: str,
+        files: dict[str, str],
+        schema: type[ConceptSchema] = ConceptSchema,
+    ):
+        self.files = files
+        self.schema = schema
+        self.extra_args = {
+            "resource_type": "concept",
+            "owner_type": "Organization",
+            "retired": "FALSE",
+            "owner_id": owner_id,
+        }
+        self.output_csv_header = self._get_csv_headers()
+        self.value_set = None
+        self.value_sets = []
+
+    def _get_csv_headers(self) -> list[str]:
+        """
+        Get the CSV headers from the first file.
+
+        Returns:
+            list[str]: The CSV headers.
+
+        """
+        if not self.files:
+            return []
+        concept_class = list(self.files.keys())[0]
+        with open(self.files[concept_class]) as f:
+            reader = csv.DictReader(f)
+            processed_row = ConceptRow(
+                next(reader), self.schema, concept_class, None, **self.extra_args
+            )
+            processed_row.process_for_csv()
+            return list(processed_row.converted_row.keys())
+
+    def collect_value_sets(self, row: dict) -> None:
+        """
+        If we have a valueset, and the row datatype is not 'codes',
+        we append the valueset to the valuesets list and reset the valueset.
+
+        Args:
+            row (dict): The row data.
+
+        Returns:
+            None
+        """
+        if self.value_set and row.converted_row["datatype"].lower() != "codes":
+            self.value_sets.append(self.value_set)
+            self.value_set = row.value_set
+        if not self.value_set:
+            self.value_set = row.value_set
+
+    def process_concept_for_csv(self, write_to_path: str) -> None:
+        """
+        Process the concept data and write it to a CSV file.
+
+        Args:
+            write_to_path (str): The path to write the CSV file.
+
+        """
+        with open(write_to_path, "w") as f:
+            writer = csv.DictWriter(f, fieldnames=self.output_csv_header)
+            writer.writeheader()
+            for concept_class, data_file in self.files.items():
```

Comment:
It seems like we could refactor this class to hold a list of `ConceptRow`, separate out the data input from the data output (i.e., make them different functions)

Reply:
So what I'm basing this on is that we seem to have some duplication in both `process_concept_for_csv()` and `process_concept_for_json()`. In both cases, we iterate through a list of files, process each file into a list of rows, create a `ConceptRow` object, convert the row into it's appropriate representation and then write that to a file.

I was suggesting breaking this up a bit into a multi-step process, i.e.:

In `__init__()`:
* Iterate through a list of files
* Process each file into a list of ConceptRows
* Store the processed ConceptRows as an attribute

In `process_for_XXX()`, for each row in the stored ConceptRows:
* Convert the row to its appropriate representation (JSON or CSV)
* Write that representation to the file

Basically what this means is instead of needing to create a bunch of temporary `ConceptRow`s on each write operation, we do the read operation once, store those results, and then can produce both outputs from a single read. This trades a little memory usage (the `ConceptRow`s are retained for the lifetime of the `ConceptTerminology` class) to avoid having to read from disk twice, which given that our data set here is ~2100 rows, seems like a reasonable trade-off.

Reply:
I see what you mean. I tend to incline towards memory than IO when it comes to working with files and I think that I why that was the more intuitive approach for me.

I think your suggested approach is much neater and if you're confident that the upper limit for the input size will easily fit in memory then I will make the refactor to do it that way.

One question though: Do you think there is a use case in which someone will want both json and csv outputs? This is perhaps the biggest reason to store the processed rows in RAM so that we can write at once for both options or other formats.